cmd/clusterctl/pkg/client/cluster/upgrader.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"github.com/pkg/errors"
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/config"
+)
+
+// ProviderUpgrader defines methods for supporting provider upgrade.
+type ProviderUpgrader interface {
```

Comment:
Minor: Convention tells me to call this Planner as in Writer => Write, Reader => Read, Planner => Plan, but it's alright if you feel ProviderUpgrader is more clear.

Reply:
I will leave ProviderUpgrader because I'm going to add the Apply method in another PR (Upgrader = Plan + Apply)